assembly/src/parsers/ast/ast.rs

```diff
@@ -0,0 +1,256 @@
+    U32CheckedXor,
+    U32CheckedNot,
+    U32CheckedShr,
+    U32CheckedShrImm(Felt),
+    U32UncheckedShr,
+    U32UncheckedShrImm(Felt),
+    U32CheckedShl,
+    U32CheckedShlImm(Felt),
+    U32UncheckedShl,
+    U32UncheckedShlImm(Felt),
+    U32CheckedRotr,
+    U32CheckedRotrImm(Felt),
+    U32UncheckedRotr,
+    U32UncheckedRotrImm(Felt),
+    U32CheckedRotl,
+    U32CheckedRotlImm(Felt),
+    U32UncheckedRotl,
+    U32UncheckedRotlImm(Felt),
+    U32CheckedEq,
+    U32CheckedEqImm(Felt),
+    U32CheckedNeq,
+    U32CheckedNeqImm(Felt),
+    U32CheckedLt,
+    U32UncheckedLt,
+    U32CheckedLte,
+    U32UncheckedLte,
+    U32CheckedGt,
+    U32UncheckedGt,
+    U32CheckedGte,
+    U32UncheckedGte,
+    U32CheckedMin,
+    U32UncheckedMin,
+    U32CheckedMax,
+    U32UncheckedMax,
+
+    // ----- stack manipulation ---------------------------------------------------------------
+    Drop,
+    DropW,
+    PadW,
+    Dup0,
+    Dup1,
+    Dup2,
+    Dup3,
+    Dup4,
+    Dup5,
+    Dup6,
+    Dup7,
+    Dup8,
+    Dup9,
+    Dup10,
+    Dup11,
+    Dup12,
+    Dup13,
+    Dup14,
+    Dup15,
+    DupW0,
+    DupW1,
+    DupW2,
+    DupW3,
+    Swap,
+    Swap2,
+    Swap3,
+    Swap4,
+    Swap5,
+    Swap6,
+    Swap7,
+    Swap8,
+    Swap9,
+    Swap10,
+    Swap11,
+    Swap12,
+    Swap13,
+    Swap14,
+    Swap15,
+    SwapW,
+    SwapW1,
+    SwapW2,
+    SwapW3,
+    SwapDW,
+    MovUp2,
+    MovUp3,
+    MovUp4,
+    MovUp5,
+    MovUp6,
+    MovUp7,
+    MovUp8,
+    MovUp9,
+    MovUp10,
+    MovUp11,
+    MovUp12,
+    MovUp13,
+    MovUp14,
+    MovUp15,
+    MovUpW2,
+    MovUpW3,
+    MovDn2,
+    MovDn3,
+    MovDn4,
+    MovDn5,
+    MovDn6,
+    MovDn7,
+    MovDn8,
+    MovDn9,
+    MovDn10,
+    MovDn11,
+    MovDn12,
+    MovDn13,
+    MovDn14,
+    MovDn15,
+    MovDnW2,
+    MovDnW3,
+    CSwap,
+    CSwapW,
+    CDrop,
+    CDropW,
+
+    // ----- input / output operations --------------------------------------------------------
+    PushAdv(Felt),
+    PushEnvLocaddr(Felt),
+    PushEnvSdepth,
+    PushLocal(Felt),
+    PushMem,
+    PushMemImm(Felt),
+    PushWLocalImm(Felt),
+    PushWMem,
+    PushWMemImm(Felt),
+    PushConstants(Vec<Felt>),
+    PopLocalImm(Felt),
+    PopMem,
+    PopMemImm(Felt),
+    PopWMem,
+    PopWMemImm(Felt),
+    PopWLocalImm(Felt),
+    LoadWAdv,
+    LoadWMem,
+    LoadWMemImm(Felt),
+    LoadWLocalImm(Felt),
+    StoreWMem,
+    StoreWMemImm(Felt),
+    StoreMemImm(Felt),
+    StoreWLocalImm(u64),
+    AdvU64Div,
+
+    // ----- cryptographic operations ---------------------------------------------------------
+    RPHash,
+    RPPerm,
+    MTreeGet,
+    MTreeSet,
+    MTreeCWM
+}
+
+// A node in a AST that can represent a block, instruction or a control flow.
+pub enum Node {
+    Instruction(Instruction),
+    IfElse(Vec<Node>, Option<Vec<Node>>),
+    Repeat(usize, Vec<Node>),
+    Call(String),
+    Exec(String),
+    Use(String),
+    While(Vec<Node>),
+    Span(Vec<Node>)
```

Comment:
What is the purpose of the `Span` node? I don't think there is such a construct at the assembly level.

Reply:
it's just a collection of instructions outside of the specific construct. It's easier to have every possibility of nodes this way, but can figure out how to remove this

Reply:
In which situations do we need such a collection? It seems to me that the only place where we might need it is for program/procedure bodies, and we have these explicitly covered with `body: Vec<Instruction>` field.

Reply:
fixed